Remove BIP70 support

fanquake opened the pull request:
This removes [BIP70](https://github.com/bitcoin/bips/blob/master/bip-0070.mediawiki) support. It also removes OpenSSL linking from Qt and building OpenSSLs `lib_ssl` in depends, as well as SSL lib detection from the build system. It's something that I'd optimistically like to do for `0.20.0`.

fanquake committed:
build: remove protobuf from depends

fanquake committed:
docs: remove protobuf from docs

fanquake committed:
Remove BIP70 Support

fanquake committed:
build: remove EVP_MD_CTX_new detection

This was added in #9475 to fix LibreSSL compatibility for
BIP70, so is no longer required.

fanquake committed:
gui: Update BIP70 support message

fanquake committed:
build: remove OpenSSL from Qt build

More info available from:
https://doc.qt.io/qt-5/ssl.html#enabling-and-disabling-ssl-support

fanquake committed:
build: remove SSL lib detection

emilengler commented:
Not tested yet but a `grep -rnI bip70` only finds matches in release notes so it looks good to me

MarcoFalke commented:
Concept ACK

![444](https://user-images.githubusercontent.com/6399679/66960086-c034d200-f039-11e9-95b4-f5c665729c4d.gif)

MarcoFalke commented:
Can the ssl libs be removed from the msvc build in `build_msvc/README.md` as well?

fanquake commented:
> Can the ssl libs be removed from the msvc build in build_msvc/README.md as well?

@sipsorcery Any chance you can take a look here, and put together some changes I can cherry-pick in?

laanwj commented:
Concept ACK

> Looks like there are still some extra SSL related symbols sneaking into the binaries for the Windows and Linux builds

That's strange! How is this possible if it doesn't link to OpenSSL? Can you list which ones?

practicalswift commented:
Concept ACK

`+22 −2,079` - very nice to see! :)

hebasto commented:
Concept ACK

fanquake commented:
> That's strange! How is this possible if it doesn't link to OpenSSL? Can you list which ones?

@laanwj The following gists contain the output of `nm -C src/qt/bitcoin-qt | grep -i ssl` after doing a depends build with this PR. I have cleaned up the output somewhat (check the revision for the symbols I've trimmed). A bunch of these will also disappear once we remove the `network` module from the Qt build.

[Windows](https://gist.github.com/fanquake/50aa19856d39913c85023b2e8dc021ae) [Linux](https://gist.github.com/fanquake/f7bc9a1e740f005e73ae0edc8f99a250) [macOS](https://gist.github.com/fanquake/1d8fbc335056e7d077e3d2bf22f168b8)

MarcoFalke commented:
(needs rebase for the macOS bip70 run in `./ci/`)

fanquake committed:
build: remove BIP70 entries from macOS Info.plist

MarcoFalke commented:
> The following gists contain the output of nm -C src/qt/bitcoin-qt | grep -i ssl after doing a depends build with this PR

None of those symbols are in `bitcoind`, right?

laanwj commented:
> @laanwj The following gists contain the output of nm -C src/qt/bitcoin-qt | grep -i ssl after doing a depends build with this PR.

Huh. I'm really confused. Can we be sure that Qt isn't doing an internal build of OpenSSL? (e.g. like the fallback it has for image format parsers)

Or is it *gasp* linking to the system OpenSSL?

fanquake committed:
gui: remove payment request file handling from OpenURI dialog

MarcoFalke commented:
Not sure why the mac build is missing

fanquake commented:
> None of those symbols are in bitcoind, right?

Some of them are, because we still using OpenSSL crypto in `bitcoind`.

> Huh. I'm really confused. Can we be sure that Qt isn't doing an internal build of OpenSSL? (e.g. like the fallback it has for image format parsers)

I don't think it is. I'm also working on slimming down our `OpenSSL` build in a [branch on top of these changes](https://github.com/fanquake/bitcoin/tree/remove_bip70_stripped_openssl). Which should make it a lot clearer what is being built and included.

theuni commented on configure.ac:
```diff
@@ -226,13 +226,6 @@ AC_ARG_ENABLE([zmq],
   [disable ZMQ notifications])],
   [use_zmq=$enableval],
   [use_zmq=yes])
-AC_ARG_ENABLE([bip70],
```
I think we probably want to leave this here, but turn ```--enable-bip70``` into an error. That way it's obvious to those who are currently manually overriding.

fanquake replied:
Have reverted this removal, and now throw an error for `--enable-bip70`.

theuni commented on src/qt/optionsmodel.cpp:
```diff
@@ -483,24 +483,6 @@ void OptionsModel::setDisplayUnit(const QVariant &value)
     }
 }
 
-bool OptionsModel::getProxySettings(QNetworkProxy& proxy) const
```
I guess the paymentprotocol was the only user of the QNetwork* ?

theuni commented on src/qt/paymentserver.h:
```diff
@@ -50,15 +43,9 @@ QT_BEGIN_NAMESPACE
-// BIP70 max payment request size in bytes (DoS protection)
-static const qint64 BIP70_MAX_PAYMENTREQUEST_SIZE = 50000;
-
 class PaymentServer : public QObject
```
I imagine the necessary scraps can probably move out of here and we can nuke PaymentServer?

fanquake replied:
Yes, that, and removing the network module from the Qt build can all be done as followups.

theuni commented on src/qt/sendcoinsentry.cpp:
```diff
@@ -2,10 +2,6 @@
 // Distributed under the MIT software license, see the accompanying
 // file COPYING or http://www.opensource.org/licenses/mit-license.php.
 
-#if defined(HAVE_CONFIG_H)
```
Please leave this in case someone uses another define down the road.

theuni commented on src/qt/transactiondesc.cpp:
```diff
@@ -2,10 +2,6 @@
 // Distributed under the MIT software license, see the accompanying
 // file COPYING or http://www.opensource.org/licenses/mit-license.php.
 
-#ifdef HAVE_CONFIG_H
```
Same, please leave these.

theuni reviewed (commented):
Concept ACK. Only reviewed very briefly.

fanquake committed:
build: skip building OpenSSL lib_ssl

fanquake commented:
Rebased on master and #17191. Reverted the `HAVE_CONFIG_H` deletions and swapped to throwing an error on `--enable-bip70` rather than dropping the check entirely as suggested by theuni.

Also re-ordered the commits so the removal steps are a bit more logical and added another change so that we are now only building OpenSSLs `lib_crypto` (no longer building `lib_ssl`).

fanquake committed:
compat: remove bswap_* check on macOS

This was originally added in #9366 to fix the gui build, as
Protobuf would also define these macros. Now that we're no-longer
using Protobuf, remove the additional check.

laanwj commented:
Code review ACK 609042a6bbd6a4526cd9713a221f827035db9935

laanwj commented:
Good, this time the OSX binaries are included in the gitian build.

jtimon commented:
Oh, wow, hadn't seen this. Yes, please. Concept ACK

fjahr commented on src/qt/transactiondesc.cpp:
```diff
@@ -48,7 +48,6 @@ QString TransactionDesc::FormatTxStatus(const interfaces::WalletTx& wtx, const i
-#ifndef ENABLE_BIP70
 // Takes an encoded PaymentRequest as a string and tries to find the Common Name of the X.509 certificate
 // used to sign the PaymentRequest.
 bool GetPaymentRequestMerchant(const std::string& pr, QString& merchant)
```
Could this method not rather be removed as well, along with the one place where it is used (line 295, same file)? I guess you try to keep the PR small but I would probably still remove this.

MarcoFalke replied:
I think this is still needed to display the merchant name for existing txs made with bip 70?

laanwj replied:
It's not the intent to remove this any time soon. Maybe in the far future on a wallet format change. This was added as *fallback* for removing of BIP70 in #16852, so that displaying of old transactions doesn't suffer.

fjahr replied:
Ok, thanks for clarifying!

fjahr reviewed (approved):
ACK 8c6081a

Reviewed code, built and ran tests. Two nits but this works.

MarcoFalke commented:
ACK 8c6081a884cd0969160955ce8687d4d4ed074db3

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

ACK 8c6081a884cd0969160955ce8687d4d4ed074db3
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUjKSgv/exTwf/JSsLIdKgCml6kmBoF0kadbbWndiUR7azT5wlTQcWV/+c9McOxO
clD0f1Sci5Fks1Dy9N5SIEoVpjh29CYSInYnbu3rQTagb4+DYi1kVtVgp0j7E1lD
eHxTD4VJIbz1rdnseGs6xJA5KU8bkZekDrwIO4u7qRZWP4ikK2dwwf96MoMt42KI
x7w6apQaSkyBjAxTVa665xOpV+PugO0CcdbMudVq3v8awC1ITrcRUH3Fp2egnuJn
ithoI7FONpQH9+ipFxQ3KO0shJWYk45nw0EMATjBZO+Amhzif1gTJacbeozC6sZs
u4jplY3sHH2pYlPdIhs3FAaHO1Cia/lwfV+k2GvXD/2Ym5qGX63ZhPYo483wghtO
3av1EsgdHdxx62TWgQChPi/5lC0DU/h5M+vnHQYE/bhzvViUrqPzyODG1NoBhKIE
JoX5SWQE7XmvCrsjhbMS/91YevGUe61AFjUwx4ITyr3zh8VGH43YOc9BZRDOyvSC
ZUO0vK0l
=K4mK
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `962930a6f6f662ed694e7153800aef1dfae989ffe2d302c9f2c6c70952527e09  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e892940108962930a6f6f662ed694e7153800aef1dfae989ffe2d302c9f2c6c70952527e09f0105a0d86f7b3cb950c950039f3ee40a16b08fff01034e4e3a6345d705a9708c7827a2574c808f1045db3625df00801f2e6306b1d11b10083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff01066f97a63ea4689441af788924211c54c08f1045db3625df008b41f74be874e98f30083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6df010dfb2fe134370524e5e7edc1cd05a34fe08f1045db3625df008b7ed3396cba19b010083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6d)

</details>

fanquake commented:
This has a release note in https://github.com/bitcoin-core/bitcoin-devwiki/wiki/0.20.0-Release-Notes-Draft.
